pkg/extensions/scheduler/task-scheduler.go

```diff
@@ -0,0 +1,94 @@
+package scheduler
+
+import (
+	"context"
+	goSync "sync"
+	"time"
+)
+
+type Status int
+
+type Cb func(Status) error
+
+const (
+	Failure Status = iota
+	Success
+	Running
+)
+
+type ITask interface {
+	DoWork(wgScheduler *goSync.WaitGroup) (Status, Cb)
+	Cancel()
+}
+
+type AbstractTask struct {
+	Task ITask
+}
+
+type TaskScheduler struct {
+	tasksQ chan ITask
+	tasksR []ITask
+	wg     *goSync.WaitGroup
+}
+
+const rateLimiterScheduler = 1000
+
+func NewTaskScheduler() *TaskScheduler {
+	ch := make(chan ITask, rateLimiterScheduler)
+
+	return &TaskScheduler{
+		tasksQ: ch,
+		wg:     new(goSync.WaitGroup),
+	}
+}
+
+const rateLimit = 1 * time.Second
+
+func (scheduler *TaskScheduler) RunScheduler(ctx context.Context) {
+	throttle := time.NewTicker(rateLimit).C
+
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				scheduler.StopScheduler()
+
+				return
+			case task := <-scheduler.tasksQ:
+				scheduler.wg.Add(1)
+				status, callback := task.DoWork(scheduler.wg)
+
+				if callback != nil {
+					_ = callback(status)
+				}
+
+				if status == Running {
```

Comment:
Why is this needed? Under what condition is the status still Running at this line? Could it be any other value?

When is a task moved from running to a different status? I think that logic should be encapsulated in this package.